### DISCUSSION TITLE
Alternate scratch/ location?

### DISCUSSION THREAD

**Discussion by stuvet** [ropensci/targets#889] [Ideas]
Running a big pipeline may fill up the scratch directory & filesystem with very large objects & crash the pipeline (even cloud-backed, clearing it beforehand & with `memory=transient`). It would be great to have the option to specify an alternative location for `_targets/scratch`. There is already a `store` option in `tar_config_set`, but you may want the rest of the metadata to remain in the pipeline directory to be tracked by `git`.

It would be great to be able to specify a separate `scratch` directory in `tar_config_set`. Personally I'd like to use a cloud bucket mounted with `gcsfuse` for the `scratch` directory, and keep the rest of the `_targets` directory on faster storage.

I'm not aware of how much work this would be, so I'm just offering this as a tentative suggestion. Workarounds would be to mount a separate drive at the `_targets/scratch` path or perhaps to make `_targets/scratch` a symlink.

Thanks

**Comment by wlandau** [comment]
Are a lot of files staying in scratch? Except for `tar_target(..., repository = "aws", format = "file")`, `targets` is supposed to automatically delete those files. You almost never need to keep them, and `tar_destroy(destroy = "scratch")` removes them.

**Reply by stuvet** [reply]
No, they're being removed as expected. Just lots of large targets coming out of `stantargets` mcmc - easily surpasses 70Gb, plus another ~15Gb files in the `.future` directory during the run. The main problem was not realising quite how much scratch storage I might need for the task as its hard to predict until you look during a pipeline run. Perhaps this could be a target for future reporting/warnings - estimated peak scratch directory size based on the completed targets stored so far? Wouldn't help for the first run, but may be useful for troubleshooting after.

Of course I could expand the drive, but expanding a machine image  on Google cloud is a one-way thing - can't be shrunk again, so seems a shame.

I'm happy with my workaround - delete the scratch directory at the start of `_targets.R` then create it again as a soft link to a `gcsfuse` mounted bucket - works well & near-infinitely expandable.

**Reply by stuvet** [reply]
Hmm, maybe I was a little premature - I thought my workaround had solved the problem but it appears that the symlink I create gets deleted again & the files saved on the local disk. I only thought it worked while I was doing partial runs of the pipeline. I guess making the whole `_targets` directory a symlink might work, though it'd probably have performance disadvantages.

**Reply by wlandau** [reply]
So I just noticed that scratch files accumulate until the very end of the pipeline and then are deleted all at once at the last minute. This is unnecessary, it would be better to delete the file in `_targets/scratch/` immediately. I am working on a solution for the AWS and GCP backends.

**Reply by wlandau** [reply]
Haven't tested this, but a temporary workaround might be to set `memory = "transient"` in `tar_target()` or `tar_option_set()`.

**Reply by stuvet** [reply]
I just realised my previous comment was incorrect - the symlink of `scratch` does work, unless you (me) make a mistake in the condition which only allows the symlink to be formed when running in the cloud. My fault. It's a good workaround.

I didn't realise that it wasn't the expected behaviour for files to hang around for a while during the pipeline run. I'd assumed they were being cleared up once that group of targets were complete. It'd be great to have a solution. Thanks!

**Reply by stuvet** [reply]
> Haven't tested this, but a temporary workaround might be to set `memory = "transient"` in `tar_target()` or `tar_option_set()`.

I already had `memory = "transient"` at the top of the file, & hadn't changed it for the problem targets. The symlink works, but sounds like you may have a different solution in the works. Thanks.

**Reply by wlandau** [reply]
@stuvet, I think I fixed the buildup of scratch/staging files in the latest commit, and I moved the cloud scratch to `file.path(tempdir(), "scratch")`. Please let me know if you still have storage issues.

**Reply by stuvet** [reply]
@wlandau Tested #893 overnight on the full pipeline with no messing around with `_targets/scratch` locations. No storage issues where they happened reliably with full pipeline runs before.

Thanks!
